### DISCUSSION TITLE
Tera template, macros.

### DISCUSSION THREAD

**Discussion by APenDel** [rwf2/Rocket#1743] [Questions] [answered]
In this case we need make a template named 'custom_macro_use'.
```rust
let context = context();
Template::render("custom_macro_use", &context)
```
custom_macro_use.html.tera
```tera
{{ macros::example_macro(arg) }}
```

#### Does exist any implementation of this?
```rust
let context = context();
Template::render("macros::example_macro", &context)
```
Thanks.

**Comment by SergioBenitez** [comment, marked as answer]
I'm sorry, I don't understand the question. Can you try expanding?

**Reply by APenDel** [reply]
Example:

In HTML, have 1 table.

```tera
// # index.html.tera
{%- import "macros" as macros -%}
....
<table>
  {%- for item in item_list -%}
    {{ macros::example_tr(item=item) }}
  {%- endfor -%}
</table>
.....

// # macros.html.tera

{%- macro example_tr(item) -%}
  <tr>
    <td name="name">{{item.name}}</td>
    <td name="desct"></td>
    <td name="price"></td>
  </tr>
{%- endmacro example_tr -%}
```

when we need a 'unic' response of macro element, we need do this:

new file 'item_of_table.html.tera'
```tera
// in this file:
{%- import "macros" as macros -%}
{{ macros::example_tr(item=item) }}
```

'POST' request:
```rust
#[post("/add_new_item", data = "<fm>")]
pub fn req_new_item(fm: Form<ById::<i64, String, ....., ......>>) -> Template {
  let mut context = HashMap::new();
  // add item...
  return Template::render("item_of_table", &context);
}
```
there is something like this?
```rust
#[post("/add_new_item", data = "<fm>")]
pub fn req_new_item(fm: Form<ById::<i64, String, ....., ......>>) -> Template {
  let mut context = HashMap::new();
  // add item...
  return Template::render("macros::example_tr", &context);
}
```
with this, we don't need 'item_of_table.html.tera'

**Reply by SergioBenitez** [reply]
If I understand your question, you're wondering if it's possible to render a Tera macro directly or if it's always necessary to render a Tera template that calls the macro. My understanding of Tera is that it is not possible. That is, you must render a template. A macro is not a template. Thus, this can't be done.

You may, however, be able to use includes and/or inheritance instead of a macro. Since the child would be a template itself, you'd be able to render it separately. 

Nevertheless, Tera is a separate project from Rocket. Whatever you can or can't do with Tera, for the most part, is just passed on.
